poseidon2-starky/src/generation.rs

```diff
@@ -1,16 +1,37 @@
-use crate::columns::{COL_INPUT_START, COL_OUTPUT_START, NUM_COLS, STATE_SIZE};
+use crate::columns::{
+    COL_1ST_FULLROUND_STATE_START, COL_2ND_FULLROUND_STATE_START, COL_INPUT_START,
+    COL_OUTPUT_START, COL_PARTIAL_ROUND_END_STATE_START, COL_PARTIAL_ROUND_STATE_START, NUM_COLS,
+    ROUNDS_F, ROUNDS_P, STATE_SIZE,
+};
+use ark_ff::{BigInteger, PrimeField};
 use num::bigint::BigUint;
 use plonky2::hash::hash_types::RichField;
 use std::convert::TryInto;
-use zkhash::ark_ff::{BigInteger, PrimeField};
+use std::ops::AddAssign;
 use zkhash::fields::goldilocks::FpGoldiLocks;
 use zkhash::poseidon2::poseidon2::Poseidon2;
 use zkhash::poseidon2::poseidon2_instance_goldilocks::POSEIDON2_GOLDILOCKS_8_PARAMS;
 
-/// Represent a row of the preimage
+pub(crate) fn scalar_to_field<RF: RichField, PF: PrimeField>(scalar: &PF) -> RF {
+    RF::from_noncanonical_biguint(BigUint::from_bytes_le(&scalar.into_bigint().to_bytes_le()))
+}
+
+pub(crate) fn scalar_to_field_vec<RF: RichField, PF: PrimeField>(scalar: &[PF]) -> Vec<RF> {
+    scalar.iter().map(|s| scalar_to_field(s)).collect()
+}
+
+pub(crate) fn field_to_scalar<RF: RichField, PF: PrimeField>(field: &RF) -> PF {
+    PF::from_le_bytes_mod_order(&field.to_canonical_biguint().to_bytes_le())
+}
+
+pub(crate) fn field_to_scalar_vec<RF: RichField, PF: PrimeField>(field: &[RF]) -> Vec<PF> {
+    field.iter().map(|f| field_to_scalar(f)).collect()
+}
+
+// Represent a row of the preimage
 #[derive(Debug, Clone, Default)]
 pub struct Row<Field: RichField> {
-    preimage: [Field; STATE_SIZE],
+    pub(crate) preimage: [Field; STATE_SIZE],
```

Comment:
so here STATE_SIZE is 8 as we want to store 8 , 32-bit words?  or its based on poseidon2 paper/implementation?

Reply:
The implementation of Poseidon2 provides various settings for the Goldilocks field, including a state size range from 8 to 20. I simply chose the one with the smallest number of columns. You could design your own settings, targeting a specific security level.